File: src/main/java/org/dementhium/action/HarvestingAction.java
package org.dementhium.action;

import org.dementhium.model.Item;
import org.dementhium.model.mask.Animation;
import org.dementhium.model.player.Player;

/**
 *
 * @author 'Mystic Flow <Steven@rune-server.org>
 *
 */
public abstract class HarvestingAction extends Action {

	public static final int TOOL_LEVEL = 0, NO_TOOL = 1, OBJECT_LEVEL = 2, HARVESTED_ITEM = 3;
	
	public interface HarvestTool {
		
		public Animation getAnimation();
		
		public int getRequiredLevel();
	}
	
	public interface HarvestObject {
		
		public int getRequiredLevel();
		
		public double getExperience();
		
	}
	
	private int cycles;
	
	private int animationCycles;
	
	private boolean started;
	
	public HarvestingAction() {
		super(1);
	}

	@Override
	public void execute() {
		HarvestObject object = getGameObject();
		if(object == null) {
			stop();
			return;
		}
		HarvestTool tool = getTool();
		Player player = getMob().getPlayer();
		int level = player.getSkills().getLevel(getSkill());
		if(++animationCycles >= 3 && tool != null) {
			player.animate(tool.getAnimation());
		}
		if(cycles-- > 0) {
			return;
		}
		if(object.getRequiredLevel() > level) {
			player.sendMessage(getMessage(OBJECT_LEVEL));
			stop();
			return;
		}
		if(tool != null) {
			if(tool.getRequiredLevel() > level) {
				player.sendMessage(getMessage(TOOL_LEVEL));
				stop();
				return;
			}
			cycles = getCycleTime();
			if(!started) {
				player.sendMessage(getStartMessage());
				player.animate(tool.getAnimation());
				started = true;
				return;
			}
			Item reward = getReward();
			if(!player.getInventory().addItem(reward.getId(), reward.getAmount())) {
				stop();
				return;
			}
			player.getSkills().addXp(getSkill(), object.getExperience());
			player.sendMessage(getMessage(HARVESTED_ITEM));
		} else {
			player.sendMessage(getMessage(NO_TOOL));
			stop();
			return;
		}
	}
	
	public abstract Item getReward();

	public abstract String getStartMessage();
	
	public abstract String getMessage(int type);
	
	public abstract int getSkill();
	
	public abstract int getCycleTime();
	
	public abstract HarvestTool getTool();

	public abstract HarvestObject getGameObject();
	
}

File: src/main/java/org/dementhium/action/HarvestingAction.kt
package org.dementhium.action

import org.dementhium.model.Item
import org.dementhium.model.mask.Animation
import org.dementhium.model.player.Player
import org.dementhium.model.player.skills.SkillId

abstract class HarvestingAction : Action(0) {

    interface HarvestTool {
        val animation: Animation get() = Animation.RESET
        val requiredLevel: Int
        val requiredSkill: Int
        val toolId: Int

        fun playerMeetsRequirements(player: Player): Boolean {
            val playerHasTool = player.inventory.contains(toolId) || player.equipment.contains(toolId)
            val playerHasLevel = player.skills.getLevel(requiredSkill) >= requiredLevel
            return playerHasTool && playerHasLevel
        }
    }

    interface HarvestObject {
        val requiredLevel: Int
        val experience: Double
    }

    var cyclesToHarvest: Int = 0

    var animationCycles = 0

    var started = false

    override fun execute() {
        val player = mob.player
        val skill = player.skills.getSkill(skillId)
        if (++animationCycles >= 3) {
            player.animate(tool.animation)
        }
        if (cyclesToHarvest-- > 0) {
            return
        }
        if (harvestObject.requiredLevel > skill.currentLevel) {
            player.sendMessage(getMessage(OBJECT_LEVEL))
            stop()
            return
        }
        val playerHasTool = tool.playerMeetsRequirements(player)
        if (playerHasTool) {
            if (tool.requiredLevel > skill.currentLevel) {
                player.sendMessage(getMessage(TOOL_LEVEL))
                stop()
                return
            }
            cyclesToHarvest = cycleTime
            if (!started) {
                started = true
                player.sendMessage(startMessage)
                player.animate(tool.animation)
                return
            }
            if (!player.inventory.addItem(harvestReward.id, harvestReward.amount)) {
                stop()
                return
            }
            skill.experience += harvestObject.experience
            player.sendMessage(getMessage(HARVESTED_ITEM))
        } else {
            player.sendMessage(getMessage(NO_TOOL))
            stop()
            return
        }
    }

    abstract val harvestReward: Item

    abstract val startMessage: String

    abstract fun getMessage(type: Int): String

    abstract val skill: Int

    val skillId: SkillId
        get() = SkillId.byId(skill)

    abstract val cycleTime: Int

    abstract val tool: HarvestTool

    abstract val harvestObject: HarvestObject

    companion object {
        const val TOOL_LEVEL: Int = 0
        const val NO_TOOL: Int = 1
        const val OBJECT_LEVEL: Int = 2
        const val HARVESTED_ITEM: Int = 3
    }
}